<a href="https://colab.research.google.com/github/celiaferreira/Covid19_RX/blob/master/4_1_SMOTE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 4.1 Oversampling

A técnica de oversampling SMOTE (Sythetic Minority Oversampling TEchnique) consiste em gerar casos sintéticos para as classes de interesse a partir dos casos já existentes.
Os novos casos são gerados na vizinhança de cada caso da classe minoritária com o intuito de se crescer o espaço de decisão desta classe e aumentar o poder de generalização dos classificadores obtidos.

Visualmente, no espaço amostral do conjunto de dados, os novos casos sintéticos são interpolados aleatoriamente ao longo do segmento de reta que une cada caso da classe minoritária a um de seus k vizinhos mais próximos, escolhidos de forma aleatória. 


Técnica exige que dados de input estejam num vetor:

In [ ]:
X_train_vec=[]
X_trainReshaped = X_train.reshape(len(X_train),200,200)
print(X_trainReshaped.shape)
for i in X_trainReshaped:
    img = i
    #print(img.shape)
    cols,colors = img.shape # gives dimensions for RGB array
    img_size = cols*colors
    img_1D_vector = img.reshape(img_size)
    # you can recover the orginal image with:
    img2 = img_1D_vector.reshape(cols,colors)
    X_train_vec.append(img_1D_vector)

(3346, 200, 200)


### 4.1.1 Target com 4 labels

In [ ]:
oversample = SMOTE(random_state=42)
# fit and apply the transform
X_over4, y_over4 = oversample.fit_resample(X_train_vec, y_train)
print(Counter(y_over4))

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


Counter({3: 1372, 1: 1372, 0: 1372, 2: 1372})


* Voltar ao shape inicial do X_train:

In [ ]:
X_over4 = X_over4.reshape(len(y_over4),200,200,1)
print('X_oversampled_shape:',X_over4.shape)
print('X_train_shape:',X_train.shape)
print('X_val_shape:',X_val.shape)
print('X_test_shape:',X_test.shape)

X_oversampled_shape: (5488, 200, 200, 1)
X_train_shape: (3346, 200, 200, 1)
X_val_shape: (1434, 200, 200, 1)
X_test_shape: (2049, 200, 200, 1)


* Confirmar nova dimensão dos dados:

In [ ]:
def count_labels(y_all):
  how_many_0 = len(np.where(y_all==0)[0])
  how_many_1 = len(np.where(y_all==1)[0])
  how_many_2 = len(np.where(y_all==2)[0])
  how_many_3 = len(np.where(y_all==3)[0])

  print('#covid:',how_many_0)
  print('#viral:',how_many_1)
  print('#bacteriana:',how_many_2)
  print('#normal:',how_many_3)

In [ ]:
count_labels(y_train)

#covid: 469
#viral: 1039
#bacteriana: 466
#normal: 1372


In [ ]:
count_labels(y_over4)

#covid: 1372
#viral: 1372
#bacteriana: 1372
#normal: 1372


- Constata-se que, via SMOTE, a classe maioritária no conjunto de treino (normal), com 1.372 imagens, serviu de referência para o oversampling, tendo sido geradas imagens até essa dimensão em cada uma das labels.

- Os casos COVID no conjunto de treino triplicaram: de 469 para 1372.

- A pneumonia bacteriana beneficiou também do oversampling, aumentando de 466 casos para 1372.

### 4.1.2 Target com 3 labels

De modo a treinar um modelo apenas com 3 labels (Covid, pneumonia, normal), concretiza-se também o oversampling neste cenário.

* Criar target com 3 labels



In [ ]:
y_train3=y_train.copy()
for i in range(0,len(y_train),1):
  if y_train[i]==2: y_train3[i]=1
  elif y_train[i]==3: y_train3[i]=2
unique_labels(y_train3)

array([0, 1, 2])

In [ ]:
y_test3=y_test.copy()
for i in range(0,len(y_test),1):
  if y_test[i]==2: y_test3[i]=1
  elif y_test[i]==3: y_test3[i]=2
unique_labels(y_test3)

array([0, 1, 2])

In [ ]:
y_val3=y_val.copy()
for i in range(0,len(y_val),1):
  if y_val[i]==2: y_val3[i]=1
  elif y_val[i]==3: y_val3[i]=2
unique_labels(y_val3)

array([0, 1, 2])

In [ ]:
oversample = SMOTE(random_state=42)
# fit and apply the transform
X_over3, y_over3 = oversample.fit_resample(X_train_vec, y_train3)
print(Counter(y_over3))

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


Counter({2: 1505, 1: 1505, 0: 1505})


* Voltar ao shape inicial do X_train:

In [ ]:
X_over3 = X_over3.reshape(len(y_over3),200,200,1)
print('X_oversampled_shape:',X_over3.shape)
print('X_train_shape:',X_train.shape)
print('X_val_shape:',X_val.shape)
print('X_test_shape:',X_test.shape)

X_oversampled_shape: (4515, 200, 200, 1)
X_train_shape: (3346, 200, 200, 1)
X_val_shape: (1434, 200, 200, 1)
X_test_shape: (2049, 200, 200, 1)


* Confirmar nova dimensão dos dados:

In [ ]:
def count_labels3(y_all):
  how_many_0 = len(np.where(y_all==0)[0])
  how_many_1 = len(np.where(y_all==1)[0])
  how_many_2 = len(np.where(y_all==2)[0])

  print('#covid:',how_many_0)
  print('#penumonia:',how_many_1)
  print('#normal:',how_many_2)

In [ ]:
count_labels3(y_train3)

#covid: 469
#penumonia: 1505
#normal: 1372


In [ ]:
count_labels3(y_over3)

#covid: 1505
#penumonia: 1505
#normal: 1505


- Constata-se que, via SMOTE, a classe maioritária no conjunto de treino (agora pneumonia, considerando 3 labels), com 1.505 imagens, serviu de referência para o oversampling, tendo sido geradas imagens até essa dimensão em cada uma das labels.
- os casos COVID aumentaram de 469 para 1505 (triplo).